### PR TITLE
[CH-123] Support short/byte/binary/decimal/array/map/struct

### PR DESCRIPTION
**Pull request by taiyang-li**
Changelog category (leave one):
- New Feature

Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Support CH column to Spark row for complex types   

Detailed Steps: 
- [x] Calculate backing buffer length for complex types(array/map/tuple) 
- [x] Implement array writer
- [x] Implement map writer
- [x] Implement tuple writer
- [x] Implement decimal writer
- [x] Implement array reader
- [x] Implement map reader
- [x] Implement tuple reader
- [x] Implement decimal reader
- [x] Add tests for above reader and writer 
- [x] Add benchmarks for above reader and writer
- [x] Complex type transformation between substrait and CH. 
- [x] Complex type transformation between Spark and substrait in gluten. 
- [x] Improve the performance of above writers(BM_CHColumnToSparkRow_Lineitem: from 328ms to 260ms)
- [x] Improve the performance of above readers(BM_SparkRowToCHColumn_Lineitem: from 359ms to 305ms)
- [x] implement short/byte/binary

Benchmark baseline branch: https://github.com/bigo-sg/clickhouse/tree/baseline

### CODE REVIEW

**Comment by kyligence-git** [comment]
Can one of the admins verify this patch?

**Comment by liuneng1994** [comment]
test this please

**Comment by liuneng1994** [comment]
test this please

**Comment by liuneng1994** [comment]
test this please

File: utils/local-engine/Parser/SparkRowToCHColumn.h [hunk 1]
```diff
@@ -182,103 +198,153 @@ class SparkRowReader
...
-    int8_t getByte(int ordinal)
+    int8_t getByte(int ordinal) const
     {
         assertIndexIsValid(ordinal);
-        return *reinterpret_cast<int8_t *>(getFieldOffset(ordinal));
+        return *reinterpret_cast<const int8_t *>(getFieldOffset(ordinal));
     }
 
-    uint8_t getUnsignedByte(int ordinal)
+    uint8_t getUnsignedByte(int ordinal) const
```
Comment (lgbo-ustc) [inline utils/local-engine/Parser/SparkRowToCHColumn.h:216]: mark inline ?

**Comment by liuneng1994** [comment]
test this please

**Comment by liuneng1994** [comment]
test this please

File: utils/local-engine/Parser/SparkRowToCHColumn.h (hunk 1 above)
Comment (taiyang-li) [reply utils/local-engine/Parser/SparkRowToCHColumn.h:216]: It's inlined by compiler already

**Comment by liuneng1994** [comment]
test this please

**Comment by liuneng1994** [comment]
test this please with 402

**Comment by liuneng1994** [comment]
test this please with 402

**Comment by liuneng1994** [comment]
test this please with 402

**Comment by liuneng1994** [comment]
test this please with 402

**Comment by taiyang-li** [comment]
test this please with 402

**Comment by taiyang-li** [comment]
test this please with 402

**Comment by taiyang-li** [comment]
test this please with 402

**Comment by taiyang-li** [comment]
test this please with 402

**Comment by taiyang-li** [comment]
test this please with 402

**Review by liuneng1994 — APPROVED** [review]
LGTM

**Pull request merged**